src/communication_abstraction/src/client_handle_helper.c

```diff
@@ -0,0 +1,273 @@
+{
+    if (type != ADUC_ConnType_NotSet && g_ClientHandleType != ADUC_ConnType_NotSet)
+    {
+        return false;
+    }
+
+    g_ClientHandleType = type;
+
+    return true;
+}
+
+IOTHUB_DEVICE_CLIENT_LL_HANDLE GetDeviceClientHandle(ADUC_ClientHandle handle)
+{
+    return *(IOTHUB_DEVICE_CLIENT_LL_HANDLE*)handle;
+}
+
+IOTHUB_MODULE_CLIENT_LL_HANDLE GetModuleClientHandle(ADUC_ClientHandle handle)
+{
+    return *(IOTHUB_MODULE_CLIENT_LL_HANDLE*)handle;
+}
+
+/**
+ * 
+ */
+_Bool ClientHandle_CreateFromConnectionString(
+    ADUC_ClientHandle* iotHubClientHandle,
+    ADUC_ConnType type,
+    const char* connectionString,
+    IOTHUB_CLIENT_TRANSPORT_PROVIDER protocol)
+{
+    _Bool success = false;
+
+    if (g_ClientHandleType != ADUC_ConnType_NotSet)
+    {
+        return success;
+    }
+
+    g_ClientHandleType = type;
+
+    if (g_ClientHandleType == ADUC_ConnType_Device)
+    {
+        IOTHUB_DEVICE_CLIENT_LL_HANDLE deviceHandle =
+            IoTHubDeviceClient_LL_CreateFromConnectionString(connectionString, protocol);
+
+        *iotHubClientHandle = (ADUC_ClientHandle)malloc(sizeof(IOTHUB_DEVICE_CLIENT_LL_HANDLE));
+
+        memcpy(*iotHubClientHandle, &deviceHandle, sizeof(IOTHUB_DEVICE_CLIENT_LL_HANDLE));
+
+        success = true;
+        goto done;
+    }
+    else if (g_ClientHandleType == ADUC_ConnType_Module)
+    {
+        IOTHUB_MODULE_CLIENT_LL_HANDLE moduleHandle =
+            IoTHubModuleClient_LL_CreateFromConnectionString(connectionString, protocol);
+
+        *iotHubClientHandle = (ADUC_ClientHandle*)malloc(sizeof(IOTHUB_MODULE_CLIENT_LL_HANDLE*));
+
+        memcpy(*iotHubClientHandle, &moduleHandle, sizeof(IOTHUB_MODULE_CLIENT_LL_HANDLE));
+
+        success = true;
+        goto done;
+    }
+
+done:
+
+    if (!success)
+    {
+        free(*iotHubClientHandle);
+    }
+
+    return success;
+}
+
+IOTHUB_CLIENT_RESULT ClientHandle_SetConnectionStatusCallback(
+    ADUC_ClientHandle iotHubClientHandle,
+    IOTHUB_CLIENT_CONNECTION_STATUS_CALLBACK connectionStatusCallback,
+    void* userContextCallback)
+{
+    IOTHUB_CLIENT_RESULT result = IOTHUB_CLIENT_INVALID_ARG;
+
+    if (g_ClientHandleType == ADUC_ConnType_Device)
+    {
+        result = IoTHubDeviceClient_LL_SetConnectionStatusCallback(
+            GetDeviceClientHandle(iotHubClientHandle), connectionStatusCallback, userContextCallback);
+    }
+    else if (g_ClientHandleType == ADUC_ConnType_Module)
+    {
+        result = IoTHubModuleClient_LL_SetConnectionStatusCallback(
+            GetModuleClientHandle(iotHubClientHandle), connectionStatusCallback, userContextCallback);
+    }
+
+    return result;
+}
+
+/**
+ * 
+ */
+IOTHUB_CLIENT_RESULT ClientHandle_SendEventAsync(
+    ADUC_ClientHandle iotHubClientHandle,
+    IOTHUB_MESSAGE_HANDLE eventMessageHandle,
+    IOTHUB_CLIENT_EVENT_CONFIRMATION_CALLBACK eventConfirmationCallback,
+    void* userContextCallback)
+{
+    IOTHUB_CLIENT_RESULT result = IOTHUB_CLIENT_INVALID_ARG;
+
```

Comment:
The result might be a bit misleading here. IIUC, if someone calling this function w/o setting g_ClientHandleType first,  it'll return INVLID_ARG?
Mind log something here to make it easier to diagnose please?

Reply:
I'll ad a log for each one. The issue is we need to keep using their pass through value but they don't exactly have a "Your clientHandle is invalid" argument. Well actually they do it's just this value.

---
In reply to: [591991585](https://github.com/Azure/iot-hub-device-update/pull/12#discussion_r591991585) [](ancestors = 591991585)